### DISCUSSION TITLE
Upgrade FreshRSS theming system

### DISCUSSION THREAD

**Discussion by aledeg** [FreshRSS/FreshRSS#3345] [Ideas]
## Why I think we should upgrade it
<img src="http://i3.kym-cdn.com/photos/images/original/000/976/824/913.gif" />

1. **themes should be extensions**
    If you look at project using themes, they treat them as extensions to allow users to make their own, to share them without the need to include them in the core. I think about Firefox and Dokuwiki but I am pretty sure others do that too.
1. **themes should use parameters**
    Now, CSS supports variables and it is [really well supported by browsers ](https://www.caniuse.com/css-variables). We should use them to offer some customization without the need to create a new theme or the need to rely on CustomCSS. Some users do know how to tweak CSS but the vast majority does not. The use of variables will allow to group color properties, font properties, and other interesting properties at the top of the theme file. Thus making them more readable. Customizing a color in a theme will be as easy as overriding a variable.
1. **themes should be documented properly**
    This one is obvious but we need proper documentation for the few who want to create themes or tweak them.

I've been digging the issues to figure what are the recurring problem with our theming system and I think that if we address those 3 items, we should be able to cover all issues.
I've created a new label to identify the aforementioned issues → https://github.com/FreshRSS/FreshRSS/labels/Theme

## Steps to upgrade
<img src="https://i.kym-cdn.com/entries/icons/original/000/019/404/upgradddd.JPG" />

I am well aware that we cannot and should not upgrade in one Big-Bang commit, so here are the different steps I have in mind.

1. **move theming configuration to its own page**
   I think we need to have the theming configuration in its own configuration page. This will allow to display customizing options relative to the selected theme. At first, it will just display the theme and allow to select one. But later, we will see the custom themes added by user extensions and the specific configuration.
1. **create theme extension**
   This will be an abstract class extending the current abstract extension class. It will allow later in the process to add specific methods only for themes like specific configuration. There will be a need for new hooks, for instance one to add the current theme to the theming configuration page.
1. **change themes to core extension**
   The theme will be converted to core extensions to use the extension system. At this point, everything should work exactly as before. But the internals will be heavily changed to get the theme file from the extension instead of the theme folder. That will require thorough testing.
1. **use variable in themes**
   Extract all repeated values in themes (colors, font, line) and convert them as variable. The idea here is not to expose all variables but at least to have something to work with. At this point, it will be easy to modify a theme by adding a rule in CustomCSS to override let say one color.
1. **add configuration in theme extension**
   Add methods in the abstract theme extension that allow to expose some of the variable of the theme for easy customization. This will create a custom CSS file (as the CustomCSS extension do). At this point, we could add [color pickers](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input/color) in the theming configuration page to easy customization. We could also add font customization, line height customization, and text justification customization. It should also allow to expose the variable used to ease overriding them when they are not exposed through configuration.
1. **add documentation on theming**
   It should include how to create a theme, how to customize a theme, how to override theme values outside of the theme parameters.

## Proof Of Concept
<img src="https://i.kym-cdn.com/photos/images/newsfeed/001/121/626/51f.gif" />

I've made a quick test to validate the use of variable in CSS and it was successful.
I've added or modified the following rules in `p/themes/Origin/origin.css`:
```css
:root {
    --red: #bd362f;
}
.btn-attention {
	background: #e95b57;
	/* background: linear-gradient(to bottom, #e95b57, #bd362f); */
	background: linear-gradient(to bottom, #e95b57, var(--red));
	color: #fff;
	border: 1px solid #c44742;
	text-shadow: 0px -1px 0px #666;
}
```
Then I created this custom rule in CustomCSS:
```css
:root{
    --red: black;
}
```
And it works! It's ugly, but it works!
![Screen Shot 2021-01-08 at 08 01 11](https://user-images.githubusercontent.com/3056148/104018493-dcb1b280-5187-11eb-887f-eee1f4b0e185.png)

## Final thoughts
If you think that we should go that way, I'll create separated issues for each action to follow through and start working on it.
If you think we should improve the plan, I'll adjust it before submitting it again and start the process over.
If you think the idea is not worth the hassle, I'll drop it.

Anyway, let me know what you think.

<img src="https://i.kym-cdn.com/photos/images/original/001/011/692/d8a.gif" />

## Useful links
- [A Complete Guide to Dark Mode on the Web | CSS-Tricks](https://css-tricks.com/a-complete-guide-to-dark-mode-on-the-web/)
- [Can I use... Support tables for HTML5, CSS3, etc](https://caniuse.com/?search=filters)
- [Can I use... Support tables for HTML5, CSS3, etc](https://caniuse.com/?search=prefers-color-scheme)
- [filter - CSS: Cascading Style Sheets | MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/filter)
- [filter | CSS-Tricks](https://css-tricks.com/almanac/properties/f/filter/)
- [prefers-color-scheme - CSS: Cascading Style Sheets | MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme)

**Comment by Frenzie** [comment]
Should be usable now. https://caniuse.com/css-variables

Also see something like https://github.com/AustP/Stylus.php

**Comment by Alkarex** [comment]
@aledeg This all sounds quite good to me 👍

**Comment by aledeg** [comment]
It's been a bit more than a week now. Do you guys think I should proceed?

**Reply by Frenzie** [reply]
Fine by me. ^_^

**Reply by Alkarex** [reply]
That sounds good. I suggest at the same time you make two distinct folders for core themes and third-party themes, just like we have for extensions (and if possible using an explicit path when using a theme instead of having to search on disk).

**Comment by aledeg** [comment]
@Alkarex @Frenzie The more I think about it and the more I feel the step 1 is not needed. I think I can just add a configuration button to open a slider. What do you think?

**Reply by Frenzie** [reply]
I'm not sure if I fully understood what point 1 meant. Taking the theme setting out of "Display" seems like a fairly trivial matter (except wrt thinking about how it should best be displayed to the user). I actually thought you wanted to start with point 4. :-)

**Reply by aledeg** [reply]
It's trivial, but I am not sure if I want to move that or keep it were it sits now. I am not sure we need a dedicated page for it.
I was thinking to start with step 2 :)

**Comment by aledeg** [comment]
So, I was thinking that upon install, a theme extension will write its content in the theme folder. But I couldn't write in that folder. I've changed permissions to allow writing for the group and I still can't write in it. I can write from cli but not from the install page.
Any idea on how to fix that?

**Reply by Frenzie** [reply]
I don't have an answer for you, but also see https://github.com/FreshRSS/FreshRSS/pull/2678#discussion_r348444939 and https://github.com/FreshRSS/FreshRSS/pull/2678 if you're looking at it.

**Reply by Alkarex** [reply]
It would be nice to only have `./data/` writeable by the Web server. See also `data/extensions-data/` which was a start to clean the way extensions save their data.
We also have `./extensions/` which is meant to be used to store third-party extensions (e.g. as a Docker volume that can be kept when upgrading FreshRSS), which could also be a fine place, and which can be given write-access.
As those are not exposed in the public Web server path, something like `./p/ext.php` must be used

**Comment by math-GH** [comment]
Is this function already available as a beta?

**Reply by aledeg** [reply]
Not yet. I have to put time on that and some thinking also.
